pkg/operator/ceph/object/controller.go

```diff
@@ -17,215 +17,329 @@ limitations under the License.
+		if errStatus != nil {
+			return reconcile.Result{}, errors.Wrap(errStatus, "failed to set status")
+		}
+		return reconcileResponse, err
+	}
 
-	c.clusterSpec.CephVersion = cluster.CephVersion
-	objectStores, err := c.context.RookClientset.CephV1().CephObjectStores(c.namespace).List(metav1.ListOptions{})
+	// Set Ready status, we are done reconciling
+	cephObjectStore.Status.Phase = k8sutil.ReadyStatus
+	err = opcontroller.UpdateStatus(r.client, cephObjectStore)
 	if err != nil {
-		logger.Errorf("failed to retrieve object stores to update the ceph version. %v", err)
-		return
+		return reconcile.Result{}, errors.Wrap(err, "failed to set status")
 	}
-	for _, store := range objectStores.Items {
-		logger.Infof("updating the ceph version for object store %q to %q", store.Name, c.clusterSpec.CephVersion.Image)
-		c.createOrUpdateStore(&store)
+
+	// Return and do not requeue
+	logger.Debug("done reconciling")
+	return reconcile.Result{}, nil
+
+}
+
+func (r *ReconcileCephObjectStore) reconcileCreateObjectStore(cephObjectStore *cephv1.CephObjectStore) (reconcile.Result, error) {
+	if r.cephClusterSpec.External.Enable {
+		_, err := opcontroller.ValidateCephVersionsBetweenLocalAndExternalClusters(r.context, cephObjectStore.Namespace, r.cephVersion)
 		if err != nil {
-			logger.Errorf("failed to update object store %q. %v", store.Name, err)
-		} else {
-			logger.Infof("updated object store %q to ceph version %q", store.Name, c.clusterSpec.CephVersion.Image)
+			// This handles the case where the operator is running, the external cluster has been upgraded and a CR creation is called
+			// If that's a major version upgrade we fail, if it's a minor version, we continue, it's not ideal but not critical
+			return reconcile.Result{}, errors.Wrapf(err, "refusing to run new crd")
 		}
 	}
-}
 
-func (c *ObjectStoreController) storeOwners(store *cephv1.CephObjectStore) metav1.OwnerReference {
-	// Set the object store CR as the owner
-	return metav1.OwnerReference{
-		APIVersion: fmt.Sprintf("%s/%s", ObjectStoreResource.Group, ObjectStoreResource.Version),
-		Kind:       ObjectStoreResource.Kind,
-		Name:       store.Name,
-		UID:        store.UID,
+	cfg := clusterConfig{
+		context:     r.context,
+		clusterInfo: r.clusterInfo,
+		store:       cephObjectStore,
+		rookVersion: r.cephClusterSpec.CephVersion.Image,
+		clusterSpec: r.cephClusterSpec,
+		DataPathMap: opconfig.NewStatelessDaemonDataPathMap(opconfig.RgwType, cephObjectStore.Name, cephObjectStore.Namespace, r.cephClusterSpec.DataDirHostPath),
+		client:      r.client,
+		scheme:      r.scheme,
+	}
+
+	err := cfg.createOrUpdateStore()
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to create object store %q", cephObjectStore.Name)
 	}
+
+	return reconcile.Result{}, nil
 }
 
+// ParentClusterChanged determines wether or not a CR update has been sent
+// func ParentClusterChanged(cluster cephv1.ClusterSpec, clusterInfo *daemonconfig.ClusterInfo, isUpgrade bool) {
```

Comment:
Instead of `ParentClusterChanged()`, now the object store will immediately start a reconcile whenever the cluster CR is updated, right? As more controllers are converted like this, it seems like it will be chaos in the logs when everything reconciles at the same time. 

Remind me how we're going to control the rolling upgrades? We will requeue the event and wait until mons and osds are updated?